### PR TITLE
Add help strings for the arguments of genpybind

### PR DESCRIPTION
**Pull request by schmitts**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+
+See [holder_type.h](./tests/holder_type.h) and [holder_type_test.py](./tests/holder_type_test.py).
+
+## `inline_base`
+
+See [inline_base.h](./tests/inline_base.h) and [inline_base_test.py](./tests/inline_base_test.py).
+
+## `keep_alive`
+
+To control the life time objects, `keep_alive` can be used.
```
Comment (kljohann) [inline README.md]:
```suggestion
To control the life time of objects passed to or returned from (member) functions, `keep_alive` can be used.
`keep_alive(bound, who)` indicates that `who` should be kept alive at least until `bound` is garbage collected.
An argument to `keep_alive` can be either the name of a function parameters or one of `return` or `this`, where
`return` refers to the return value of the function and `this` refers to the instance a member function is called on.
```

File: README.md [hunk 2]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+Parent(Child *child);
+```
+
+When the instance of `Parent` is deleted from Python, `child` will not be deleted as well.
+
+See [keep_alive.h](./tests/keep_alive.h) and [keep_alive_test.py](./tests/keep_alive_test.py).
+
+## `module`
+
+C++ namespaces can be turned into Python modules with the `module` keyword:
```
Comment (kljohann) [inline README.md]:
```suggestion
Using the `module` keyword C++ namespaces can be turned into submodules
of the generated Python module. In the following example, `X` would be exposed
as `name_of_module.submodule.X`, where `name_of_module` is the name of the
outer Python module.
```

File: README.md [hunk 3]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+const Nested &cref() const;
+
+GENPYBIND(return_value_policy(copy))
+Nested &ref_as_copy();
+
+GENPYBIND(return_value_policy(reference_internal))
+Nested &ref_as_ref_int();
+```
+
+`ref()` and `cref()` both return a copy. `ref_as_copy()` returns a copy although C++ returns a reference and `ref_as_ref_int()` returns a reference.
```
Comment (kljohann) [inline README.md]:
```suggestion
By default, the `automatic` [return value policy](https://pybind11.readthedocs.io/en/master/advanced/functions.html#return-value-policies)
of pybind11 is used.  In the case of `ref` and `cref` in the example this amounts to
"return by value" for the wrapped Python functions. This behavior is unchanged
when the function is explicitly annotated to return by value (see `ref_as_copy`).
As `ref_as_ref_int` demonstrates, any other return value policy supported by
pybind11 can be set. In this case `reference_internal` is used to return a reference
to an existing object, whose life time  is tied to the parent object.
```

File: README.md [hunk 4]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+```cpp
+Nested &ref();
+
+const Nested &cref() const;
+
+GENPYBIND(return_value_policy(copy))
+Nested &ref_as_copy();
+
+GENPYBIND(return_value_policy(reference_internal))
+Nested &ref_as_ref_int();
```
Comment (kljohann) [inline README.md]: Needs to be wrapped in a class for the example to make sense. Unfortunately the GitHub interface is too clumsy to make that edit. :(

File: README.md [hunk 5]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+
+```cpp
+class Unannotated {};
+
+class GENPYBIND(hidden) Hidden {};
+
+class GENPYBIND(visible) Visible {};
+```
+
+Unannotated parts are excluded by default, but the intend of hiding it
```
Comment (kljohann) [inline README.md]:
```suggestion

Anything without an annotation is excluded by default, but the intent of hiding it
```

File: README.md [hunk 6]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+from bindings can be explicitly stated by the keyword `hidden`.
+
```
Comment (kljohann) [inline README.md:392]:
```suggestion

If a namespace is annotated with `visible`, any contained entity will be made visible
by default, even if it has no `GENPYBIND` annotations. `GENPYBIND(hidden)` can then
be used to hide it.
```

File: README.md [hunk 7]
```diff
@@ -153,6 +153,241 @@ a single C++ tool based on the internal `libtooling` API.  I eventually plan to
...
+If a binding is supposed to be generated is controlled by the visibility keywords `visible` and `hidden`:
+
+```cpp
+class Unannotated {};
+
+class GENPYBIND(hidden) Hidden {};
+
+class GENPYBIND(visible) Visible {};
+```
+
```
Comment (kljohann) [inline README.md:385]:
```suggestion

Any `GENPYBIND` annotation will make the annotated entity visible. 
As a consequence `visible` can be removed from the argument list,
as soon as there are any other arguments to `GENPYBIND`.
```

**Review by kljohann — COMMENTED** [review]
This is wonderful! Ideally we would also add links/references to the documentation of  the underlying pybind11 concepts. :tada: Some minor additions and comments.

File: README.md (hunk 2 above)
Comment (schmitts) [reply README.md]: the the

Comment (kljohann) [reply README.md]: Whoops. :) Fixed. What do you think about the suggestion? Should we add it to your change?

File: README.md (hunk 4 above)
Comment (kljohann) [reply README.md]: I can also make this change during the merge, if you prefer?

File: README.md (hunk 6 above)
Comment (kljohann) [reply README.md:392]: I can also add this, if it's fine for you.

Comment (schmitts) [reply README.md:392]: sure, go ahead!

File: README.md (hunk 4 above)
Comment (schmitts) [reply README.md]: you can make the change

File: README.md (hunk 2 above)
Comment (schmitts) [reply README.md]: yep, change/add as you like

**Comment by kljohann** [comment]
@schmitts Updated. (Finally! :sweat_smile:) Can you take a look whether things look fine for you?

**Comment by kljohann** [comment]
Merged, thanks Sebastian! :bow:

**Pull request closed without merging**